Fix: Cut-Away View+Giant Screenshot results in most of the image being transparent area

CoderUndefined opened the pull request:
The previously mentioned bug occurs when the user attempts to take a giant screenshot while using the cut-away view. The actual dimensions of the giant screenshot remained the same, but only the potion is visible that is set with the cut-away view. This results in a very large blank area on the screenshot.

This behavior was changed so that when a giant screenshot is taken with cut-away view enabled, only the small  portion is taken and all surrounding blank area is removed from the giant screenshot. The z-height of the objects are taken into account.

It is achieved by limiting the viewport size of the giant viewport, using the coordinates provided by the cut-away view when it is enabled.

Screenshots can be provided if necessary, as some of them are too tall to fit in this post.

CoderUndefined commented:
Note that I am doing this as part of an assignment. The issue needs to be closed in about a month in some way or else this assignment fails.

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -339,25 +355,104 @@ static void ReleaseDPI(rct_drawpixelinfo& dpi)
+        }
+        else
+        {
+            leftTileCoords = gClipSelectionA;
+            rightTileCoords = gClipSelectionB;
+            topTileCoords = CoordsXY(gClipSelectionA.x, gClipSelectionB.y);
+            bottomTileCoords = CoordsXY(gClipSelectionB.x, gClipSelectionA.y);
+
+            rightTileCoords += CoordsXY(32, 32);
+        }
+
+        minY = gClipSelectionA.y / 32;
+        maxY = gClipSelectionB.y / 32;
+        minX = gClipSelectionA.x / 32;
+        maxX = gClipSelectionB.x / 32;
+        int height = GetTallestVisibleTileTopLimited(minX, maxX, minY, maxY, rotation);
+
+        left = translate_3d_to_2d_with_z(rotation, CoordsXYZ(leftTileCoords, 0)).x;
+        right = translate_3d_to_2d_with_z(rotation, CoordsXYZ(rightTileCoords, 0)).x;
+        top = translate_3d_to_2d_with_z(rotation, CoordsXYZ(topTileCoords, height)).y;
+        bottom = translate_3d_to_2d_with_z(rotation, CoordsXYZ(bottomTileCoords, 0)).y;
+    }
+    else
+    {
+        // Run old default code
+        leftTileCoords = GetEdgeTile(mapSize, rotation, EdgeType::LEFT, false);
+        rightTileCoords = GetEdgeTile(mapSize, rotation, EdgeType::RIGHT, false);
+        bottomTileCoords = GetEdgeTile(mapSize, rotation, EdgeType::BOTTOM, false);
+
+        // Centre the coordinates so we don't have a hard crop at the edge of the visible tile
+        leftTileCoords += CoordsXY(16, 16);
+        rightTileCoords += CoordsXY(16, 16);
+        bottomTileCoords += CoordsXY(16, 16);
+
+        // Calculate the viewport bounds
+        left = translate_3d_to_2d_with_z(rotation, CoordsXYZ(leftTileCoords, 0)).x;
+        top = GetTallestVisibleTileTop(mapSize, rotation);
+        right = translate_3d_to_2d_with_z(rotation, CoordsXYZ(rightTileCoords, 0)).x;
+        bottom = translate_3d_to_2d_with_z(rotation, CoordsXYZ(bottomTileCoords, 0)).y;
+    }
```
I think this entire function can be reverted. You only have to modify the `GetEdgeTile` function to return the correct tiles for when cut-away view selection is active.

CoderUndefined replied:
Most of the new code has now been migrated to GetEdgeTile. Note that the old code is still on an else branch, otherwise the giant screenshot fails to properly render and the image becomes very large in terms of pixel dimensions.

CoderUndefined commented on src/openrct2-ui/windows/ViewClipping.cpp:
```diff
@@ -343,6 +344,16 @@ class ViewClippingWindow final : public Window
 
     void OnOpen() override
     {
+        // only occurs for the first time when a scenario starts and when this window opens.
+        // resets the clipping values to the default values
+        if (firstStart)
+        {
+            firstStart = false;
+            int zero = 0;
+            gClipSelectionA = { zero, zero };
+            gClipSelectionB = { MAXIMUM_TILE_START_XY, MAXIMUM_TILE_START_XY };
+        }
+
```
There is an issue that occurs when the cut-away view is used on the current scenario, then a new scenario is loaded.
It previously did not reset the values. This means if the cut-away view is opened up in the new scenario, it loads the old values that were used in the previous scenario. This can be problematic if the new scenario uses a smaller map than the previous scenario.

CoderUndefined replied:
gClipSelectionA and gClipSelectionB are global variables that are defined in Paint.cpp and Paint.h respectively and seem to be used to store the boundaries for the cut-away view. I would rather handle this current pull request first. I can either open up an issue to highlight the need to refactor these global variables or open another pull request after this.

Could be a good first issue for someone else,

CoderUndefined commented:
There is currently an issue with using the new code with the normal Giant Screenshot, hence the old code on an else branch.
There are also offset numbers in the GetEdgeTile in the new code so that the image is properly cropped in the right way.

CoderUndefined commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -290,7 +291,7 @@ static int32_t GetHighestBaseClearanceZ(const CoordsXY& location)
         } while (!(element++)->IsLastForTile());
     }
-    return z;
+    return z + 16;
```
This was a mistake from my side. I changed that value at some point in an attempt to fix a cropping issue at the top of the image. This is no longer required since the cropping issue was fixed differently, this was reverted to the original state.

CoderUndefined replied:
Turns out some sprites are cropped of with the original z value. This padding has been added to the new function GetTallestVisibleTileTopLimitedm, which is added after the function call of GetHighestBaseClearanceZ
`            int32_t z = GetHighestBaseClearanceZ(location) + 16;`

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -220,61 +221,129 @@ static CoordsXY GetEdgeTile(TileCoordsXY mapSize, int32_t rotation, EdgeType edg
     int32_t upperX = (visible ? mapSize.x - 2 : mapSize.x - 1) * COORDS_XY_STEP;
     int32_t upperY = (visible ? mapSize.y - 2 : mapSize.y - 1) * COORDS_XY_STEP;
-    switch (edgeType)
+    rct_viewport* r = window_get_viewport(window_get_main());
+
+    // check if cut-away view is active, otherwise use the old code instead.
+    if ((r != nullptr && r->flags & VIEWPORT_FLAG_CLIP_VIEW)
+        && (gClipSelectionA.x != 0 || gClipSelectionA.y != 0 || gClipSelectionB.x != MAXIMUM_TILE_START_XY
+            || gClipSelectionB.y != MAXIMUM_TILE_START_XY))
```
Wouldn't it be possible to just update lower and upperX/Y to the view clip boundaries instead of having two almost identical switch blocks?

CoderUndefined replied:
Nevermind, there was a single typo.
Commiting very soon

CoderUndefined replied:
Left and right edges are moved to avoid the hard crop at the left and right edge

Broxzier replied:
You could see whether or not to use view clipping like this:

```cpp
    auto* const mainWindow = window_get_main();
    const auto* const mainViewport = window_get_viewport(mainWindow);
    const bool useViewClipping = (mainWindow != nullptr && mainViewport->flags & VIEWPORT_FLAG_CLIP_VIEW);
```

Then use that bool to determine the value for the min/max edge tiles, something like this:

```cpp
    const auto lowerX = useViewClipping ? gClipSelectionA.x : 0;
    const auto lowerY = useViewClipping ? gClipSelectionA.y : 0;
    const auto upperX = useViewClipping ? gClipSelectionB.x : ((gMapSize.x - 1) * COORDS_XY_STEP);
    const auto upperY = useViewClipping ? gClipSelectionB.y : ((gMapSize.x - 1) * COORDS_XY_STEP);
```

You can leave the rest almost untouched (only needing to change lower to lowerX and lower Y), and you are nearly there already, without the need to a lot of additional code. You then still need to make sure the returned tiles are included in their entirety, but that's shouldn't be handled differently if view clipping is on or off.

Broxzier replied:
`GetTallestVisibleTileTop` also has to take the horizontal clip area into account instead of the full map, as well as the vertical one (not done at all yet). You could loop over just the area like this:

```cpp
    TileCoordsXY startCoords = useViewClipping ? TileCoordsXY(gClipSelectionA) : TileCoordsXY{ 1, 1 };
    TileCoordsXY endCoords = useViewClipping ? TileCoordsXY(gClipSelectionB) : TileCoordsXY{ gMapSize.x - 2, gMapSize.y - 2 };

    int32_t minViewY = 0;
    for (int32_t y = startCoords.y; y <= endCoords.y; y++)
    {
        for (int32_t x = startCoords.x; x <= endCoords.x; x++)
        {
            // ...
```

CoderUndefined commented:
The function GetTallestVisibleTileTop has been replaced by the similar function GetTallestVisibleTileTopLimited because the former is no longer used and should be therefore removed to avoid a compiler warning. I might as well just rename the new function back to GetTallestVisibleTileTop. There is also an unused parameter rotation, which can be safely removed.

Broxzier commented:
To help you quickly see if cropping works right, disable the transparency option for giant screenshots in the options window.

![transparent screenshot](https://user-images.githubusercontent.com/9705046/172935998-9c97322e-0224-4760-87b5-c5c4e1d0edac.png)

CoderUndefined commented:
I am going to sleep. As it is unfinished, I marked this pull request as draft for the time being. Work will resume tomorrow.

CoderUndefined commented:
also what the heck is this? This only occurs with the large loop.
I am aware that it is composed of multiple pieces in terms of rendering but if you set the vertical clipping value in the cut-away view just right...

![Funtopia 2022-06-09 23-01-12](https://user-images.githubusercontent.com/73300809/172943917-51fea8da-1a05-4b06-b41e-71d96d253844.png)

Broxzier commented:
That's just a side effect of how track blocks work when they cover multiple tiles. Nothing related to the screenshot code.

I just realized that the rotation is not really important when determining the screen bounds, you can just simply take the 4 corners, take their screen coordinates, and base the bounds on that. That gets rid of a lot of code, and makes it simpler to understand, too.

I made a quick test branch where you can see this in action, the vertical clipping and finding the highest clearance height of the clipped area need to be worked on further. This reduces the code by 50 lines, and only improves its behaviour for view clipping: https://github.com/OpenRCT2/OpenRCT2/compare/develop...Broxzier:refactor/giant-screenshot-cleanup

I think it makes sense to go this approach for this PR too.

CoderUndefined commented:
To get the correct height for the giant screenshot when clipping it off I used this
gClipHeight needs to be mutiplied by 8 to correspond to one unit which is about 8 pixels tall.
Then at the end some padding is added so it does not crop it off.
It needs to take really tall objects into account.

`ceiling = useViewClipping ? std::min((gClipHeight)*8 + 192, ceiling + 64) : ceiling + 64;`

Broxzier commented:
The calculated z value should take the visibility of the elements into account already, no need to fall back to the active clip height. If an element is not visible, it should just not contribute to the highest point calculation. You'll need to do something similar to what the paint code does in `GetTallestVisibleTileTop`:

https://github.com/OpenRCT2/OpenRCT2/blob/88b3e88bdda85f21c4421fdee34fe6144fe4fe44/src/openrct2/paint/tile_element/Paint.TileElement.cpp#L224-L226

CoderUndefined commented:
I tested your changes and incorporated them as far as it works.

This is still not even remotely finished yet. Once it is actually finished I put this pull request as ready for review.

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -25,6 +25,7 @@
 #include "../drawing/X8DrawingEngine.h"
 #include "../localisation/Formatter.h"
 #include "../localisation/Localisation.h"
+#include "../paint/Paint.h"
```
This include is no longer needed, now with the `extern` lines below.

CoderUndefined replied:
There were other unnecessary includes so I might as well remove them too.

CoderUndefined commented:
Regarding the old code of GetTallestVisibleTileTop and anything related to it.

This mostly occurs on rotation 0, but also to a lesser degree on rotation 1 and 3, but not on 2.
For this, rotation 0 is used.

If you use the cut-away view on the most southernmost portion of the map, then take the giant screenshot, that giant screenshot is about as tall as the full giant screenshot in terms of height.
If the same is tried on rotation 1 and 3 then the resulting screenshot is about half as tall as the full giant screenshot.

I think this occurs because the northernmost tile (on rotation 0) seems to be used to calculate the height. Said northernmost tile would be the easternmost and westernmost tile on rotation 1 and 3 respectively which is why in these cases it is half as tall as the full giant screenshot.

Rough sketch below
![image](https://user-images.githubusercontent.com/73300809/173183568-0150fc78-40ce-4055-90fe-5c013fdfe3a0.png)




I have currently more important matters right now, so work on this will be paused for a week.

One last disclaimer is that I am doing this as part of an assignment. We still have about 1 month left to finish this which is plenty of time.

Broxzier commented:
I guess the offset is because the initial value of 0 is at the corner. Can probably be fixed by starting with any of the tile coords that you already know are within bounds, making it optional (until after first assignment), or initializing it with a very large number.

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -21,21 +21,13 @@
 #include "../core/Path.hpp"
-#include "../drawing/Drawing.h"
 #include "../drawing/X8DrawingEngine.h"
-#include "../localisation/Formatter.h"
```
Do not remove includes that are used, even if the file compiles fine (since another header probably includes it). Formatter is used in `screenshot_giant`, so this header should definitely remain here. This goes for many other includes that are removed now too.

CoderUndefined replied:
Weird, the IDE tells me that these are unused and can be removed safely. Also compiled just fine.
If you tell me so then I put these includes back in. Except for Paint.h

Broxzier replied:
It compiles fine because these headers are included by other headers, but by removing them here you rely on them staying in the other headers. If they are removed from these headers, you suddenly get compiler errors in this file even when nothing is changed here.

CoderUndefined commented:
Now with your GetTallestVisibleTileTop function and using a larger initial minViewY value that previous issue that occurs on rotation 0 is mostly fixed. Screenshot height still varies by a bit regarding the rotation, but not by much. 

This could be solved by adding the right amount of padding depending on the rotation but the code will not be elegant since this would rely on a switch block. Alternatively just leave as it is.

Vertical clipping in the z axis still does not cut off properly, this is being worked on.

CoderUndefined commented:
Incorporated that Paint.TileElement code into the GetHighestBaseClearanceZ function to cut anything off that is past the gClipHeight value.
The bool useViewClipping had to be used from the GetGiantViewport function.

Broxzier commented on src/openrct2-ui/windows/ViewClipping.cpp:
```diff
@@ -68,10 +68,11 @@ class ViewClippingWindow final : public Window
     CoordsXY _selectionStart;
-    CoordsXY _previousClipSelectionA;
-    CoordsXY _previousClipSelectionB;
+    CoordsXY _previousClipSelectionA = { 0, 0 };
+    CoordsXY _previousClipSelectionB = { MAXIMUM_TILE_START_XY, MAXIMUM_TILE_START_XY };
```
I still think these changes do not belong to this PR. It's a different issue, not directly related to giant screenshots.

CoderUndefined replied:
Reverted the file entirely to its original state. Once this PR is finished, that one is the next.

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -207,106 +212,40 @@ std::string screenshot_dump_png_32bpp(int32_t width, int32_t height, const void*
         }
     }
-    return minViewY - 256;
+    return minViewY - 32;
```
This is for some padding at the top?

CoderUndefined replied:
Currently, the padding is only required because of the weird collision check of the observation tower at the top. Everything else seems to work as it should, which means that the sprite will not be cropped off.

CoderUndefined replied:
to be more precise, the bounding box on the observation tower at the top appears to be lower than the actual sprite and it is possible to place things on top of the observation tower (which probably was always the case)

Broxzier replied:
Yeah, there are quite some objects, track elements, shops, and entrance styles in the game where the clearance height does not match the visuals. I guess the original 256 pixels was for this, I wonder if 32 is enough. 256 seems a bit too much, though.

Any way, I'd like this commented, to explain why the magic number is there.

CoderUndefined replied:
Comment was added

Broxzier replied:
Nitpicky, but we usually place comments above code, not behind it. Comments are only placed behind a line for lists and other situations where a comment would interrupt the code.

CoderUndefined replied:
Fixed comment placement

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -207,106 +212,40 @@ std::string screenshot_dump_png_32bpp(int32_t width, int32_t height, const void*
             auto location = TileCoordsXY(x, y).ToCoordsXY();
-            int32_t z = GetHighestBaseClearanceZ(location);
+            int32_t z = GetHighestBaseClearanceZ(location, useViewClipping);
             int32_t viewY = translate_3d_to_2d_with_z(rotation, CoordsXYZ(location, z)).y;
```
`location` has a bit of an offset, since it's from `TileCoordsXY`, it points at the corner. Might want to get the tile center here instead, and subtract 16 from y to cover the full top of the clearance for the tile. This would explain why get different offsets from different rotations. There's a helper function to get the tile center.

CoderUndefined replied:
Implemented ToTileCentre helper function, this now has fixed that offset issue regarding different rotations

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -337,23 +276,44 @@ static void ReleaseDPI(rct_drawpixelinfo& dpi)
+    auto screenCoords4 = translate_3d_to_2d_with_z(rotation, { corners[3].ToCoordsXY().ToTileCentre(), 0 });
+
+    // Current issue with height clipping is
+    // if rotation is 0 (north is north) then the image height is equal to the distance of the northernmost point of the map
+    // same applies to a lesser degree if rotation is 1 or 3
```
I think this comment is related to my previous comment.

Broxzier reviewed (changes requested):
Left a couple new remarks. I haven't tested the new build yet.

CoderUndefined commented:
Implemented most changes and reverted ViewClipping.cpp to its original state. That being said, funny things might happen if you use this new feature on a very big map, then go to a different, very small map (say, Micro Park) and try this feature again.

Broxzier commented:
Most changes look good, now there is only one new issue that has to be fixed. When all elements on the map are outside the view clipping range, it will default to the max integer. This results in absolutely gigantic black screenshots.

![image](https://user-images.githubusercontent.com/9705046/175126593-c939ee86-3921-42a2-937a-86df5474fdba.png)

It would make sense in these cases to not make a screenshot at all.

CoderUndefined commented:
This is just a workaround, do not screenshot at all if the resulting image is too large. Simply check if the image is wider than 64000 or taller than 32000 pixels.

As for new strings, this would require internationalization. Could at least do two languages for now I guess.

CoderUndefined commented:
For now, if the resulting gClipSelectionB values are too high, which causes that extremely large screenshotw when that is not supposed to happen, it simply takes a regular giant screenshot.

James103 commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -368,8 +372,7 @@ void screenshot_giant()
         {
-            context_show_error(STR_SCREENSHOT_FAILED, STR_NONE, {});
-            return;
+            throw std::exception();
```
I don't see any other `throw std::exception();` in the codebase.
Could this if block be changed into a `Guard::Assert`?

Example:
`Guard::Assert(viewport.width < 64000 && viewport.height < 32000, "Screenshot too large")`

CoderUndefined replied:
This causes a crash to desktop if Guard::Assert is used

Broxzier replied:
`std::exception` should not be thrown by itself, always use a class that derives from it. In this case `std::domain_error` seems appropriate, although I think this should be handled differently anyway. These magic numbers just happen to work because it matches the max map size, instead you should fall back to the actual map's edges if the selection covers it all.

CoderUndefined replied:
Changed this to std::domain_error. That being said, this exception should never occur at all as of now. Right now it is only exists just in case so the game does not outright crash if that very unlikely case. Could remove this exception if this never occurs at all.

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -337,23 +276,45 @@ static void ReleaseDPI(rct_drawpixelinfo& dpi)
+    const auto* const mainViewport = window_get_viewport(mainWindow);
+    bool useViewClipping = (mainViewport != nullptr && mainViewport->flags & VIEWPORT_FLAG_CLIP_VIEW);
 
+    if (gClipSelectionA.x == 0 && gClipSelectionA.y == 0 && gClipSelectionB.x >= 32000 && gClipSelectionB.y >= 32000)
+    {
+        useViewClipping = false;
+    }
```
This is not right, it disables vertical clipping too.

CoderUndefined replied:
I am using a separate bool now to check if that selection is way too large.
Then perform the check at auto corners. 
If both bools are true, then the limiteed corner selection is used. Otherwise use the entire map.

This change leaves useViewClipping untouched, meaning that vertical clipping will remain enabled.

CoderUndefined commented:
There are two constants, MAXIMUM_TILE_START_XY and MAXIMUM_MAP_SIZE_BIG. For some reason, gClipSelectionB initializes with MAXIMUM_TILE_START_XY but when resetting that selection it uses MAXIMUM_MAP_SIZE_BIG - 1;

gClipSelectionB is initialized at Paint.cpp and that selection occurs in ViewClipping.cpp. As of now, these files are untouched, nothing was edited from me.

In practice these translate to 32000 and 32031 respectively.

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -207,106 +212,41 @@ std::string screenshot_dump_png_32bpp(int32_t width, int32_t height, const void*
 {
-    int32_t minViewY = 0;
-    for (int32_t y = 1; y < mapSize.y - 1; y++)
+    int32_t minViewY = INT32_MAX;
```
```suggestion
    int32_t minViewY = std::numeric_limits<int32_t>::max();
```

CoderUndefined replied:
Done already

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -337,23 +277,47 @@ static void ReleaseDPI(rct_drawpixelinfo& dpi)
+    bool clipSelectionNotTooLarge = true;
 
+    // gMapSize must be multiplied by 32 per tile to match gClipSelection
+    if (gClipSelectionB.x - gClipSelectionA.x >= gMapSize.x * 32 || gClipSelectionB.y - gClipSelectionA.y >= gMapSize.y * 32)
+    {
+        clipSelectionNotTooLarge = false;
+    }
```
A nicer way to handle this is to just clamp the corner coords in `corners` below to be inside gMapSize.

CoderUndefined replied:
See comment below

Broxzier reviewed (commented):
Just two more things to improve. I cannot test this until Sunday evening.

Broxzier commented:
> I might as well remove that domain_error exception, or maybe keep it just in case? I still have not figured out to trigger that exception.

Once clamped, we can be sure the coordinates are valid and the screenshot won't be absurdly large. Then this can be removed entirely.

CoderUndefined commented:
Remove that if condition entire and replace cornerCoords with something like this

```    // Get the tile coordinates of each corner
    const TileCoordsXY cornerCoords[2][4] = {
        {
            // Map corners
            { 1, 1 },
            { gMapSize.x - 2, gMapSize.y - 2 },
            { 1, gMapSize.y - 2 },
            { gMapSize.x - 2, 1 },
        },
        {
            // Horizontal view clipping corners
            TileCoordsXY{ CoordsXY{ std::max(gClipSelectionA.x, 32), std::max(gClipSelectionA.y, 32) } },
            TileCoordsXY{ CoordsXY{ std::min(gClipSelectionB.x, (gMapSize.x - 2) * 32),
                                    std::min(gClipSelectionB.y, (gMapSize.y - 2) * 32) } },
            TileCoordsXY{ CoordsXY{ std::max(gClipSelectionA.x, 32), std::min(gClipSelectionB.y, (gMapSize.y - 2) * 32) } },
            TileCoordsXY{ CoordsXY{ std::min(gClipSelectionB.x, (gMapSize.x - 2) * 32), std::max(gClipSelectionA.y, 32) } },
        },
    };```

Broxzier commented on src/openrct2/interface/Screenshot.cpp:
```diff
@@ -337,23 +277,42 @@ static void ReleaseDPI(rct_drawpixelinfo& dpi)
+
+    auto left = std::min({ screenCoords1.x, screenCoords2.x, screenCoords3.x, screenCoords4.x }) - 32;
+    auto top = GetTallestVisibleTileTop(rotation, corners[0], corners[1], useViewClipping);
+    auto bottom = std::max({ screenCoords1.y, screenCoords2.y, screenCoords3.y, screenCoords4.y });
```
I was wondering why this doesn't need any subtraction since the coordinate is the center of the lowest tile, but that seems to be because the tiles at the map's edge have a base eight of 2 by default. This height offset just happens to be the same.

CoderUndefined commented:
My initial testing so far does not show any issues except for one, caused by the lack of changes in ViewClipping.cpp

Some funny things may happen if you go from a large park to a small one like Micro Park
and
leave the original coords of the cut-away view intact (which are out of bound in a very small maps)

As you said, that is a separate issue that needs a separate pull request.

CoderUndefined commented:
Created a new pull request for that separate issue which is numbered #17510 . It should ideally be used with this pull request

Broxzier commented:
Just tested this and found no further issues. Please add a changelog entry for this fix, and then this can get merged.

Broxzier commented:
Your branch is a bit out of date, can you rebase it on current develop and move the changelog entry to the appropriate list? I don't expect any other merge conflicts.

Broxzier commented:
All commits can be squashed into one, no need to keep the history for each individual change. I'll rebase this for you.

CoderUndefined commented:
Also GitHub takes a while until all checks are passed, which I had to wait just to be sure

fidwell commented:
I know I'm late to discovering this but THANK YOU so much for adding this feature!! This saves me so much time; I really appreciate it.

CoderUndefined commented:
Thanks man.
